clang-tidy-14 findings of Meson based project not shown in "files changed" tab of pull request

When adding cpp-linter to my repo I noticed that in the console output it finds all issues I added for testing, but these clang-tidy findings are not shown in the Github UI in the "files changed" tab at all. After I enabled thread-comments the findings are shown in the conversation of the pull request.

https://github.com/lurtz/sleep-proxy/pull/6

I guess the reason is using clang-tidy-14. I first used clang-tidy-12 but that had issues with finding system headers and thus provided many false positives.

> The fact that the paths start with a `/` is a serious problem because technically that is an absolute path.
> 
>     * `build/../src/blah` is a relative path with relative redirects (`..`)
> 
>     * `/build/../src/blah` is an absolute path with relative redirects (`..`)
> 
> 
> The fix I proposed above will not recognize `/build/../src/` as a relative path as you can see from the following interactive python session:
> 
> ```python
> >>> from pathlib import PurePath
> >>> PurePath("/build/../src").is_absolute()
> True
> >>> PurePath("build/../src").is_absolute()
> False
> ```

Maybe it is enough to remove the "dir/.." pattern. I got it working with the following Python code:

    In [1]: from pathlib import Path
    In [2]: Path('/foo/bar/../src').resolve()
    Out[2]: PosixPath('/foo/src')
    In [3]: Path('foo/bar/../src').resolve()
    Out[3]: PosixPath('/home/lutz/workspace/sleep-proxy/.github/workflows/foo/src')


I checked the console output of the last failing and then successful build (where I tweak the compile_commands.json so that everthing starts from the source directory).

[No annotations:](https://github.com/lurtz/sleep-proxy/actions/runs/4211802730/jobs/7310386899)

    INFO:CPP Linter:Running "clang-tidy-14 --export-fixes=clang_tidy_output.yml -p /home/runner/work/sleep-proxy/sleep-proxy/build --extra-arg= src/emulateHost.cpp"
    DEBUG:CPP Linter:Output from clang-tidy:
    ../src/emulateHost.cpp:28:4: warning: unused variable 'x' [clang-diagnostic-unused-variable]
              x
              ^
    /home/runner/work/sleep-proxy/sleep-proxy/build/../src/emulateHost.cpp:32:7: warning: redundant boolean literal in if statement condition [readability-simplify-boolean-expr]
      if (false) {
      ~~~~^~~~~~~~
    /home/runner/work/sleep-proxy/sleep-proxy/build/../src/emulateHost.cpp:33:18: warning: variable 'sptr' is not initialized [cppcoreguidelines-init-variables]
              std::string * sptr;
                            ^
    note: this fix will not be applied because it overlaps with another fix
    /home/runner/work/sleep-proxy/sleep-proxy/build/../src/emulateHost.cpp:40:15: warning: do not use pointer arithmetic [cppcoreguidelines-pro-bounds-pointer-arithmetic]
        setup_log(argv[0], 0, LOG_DAEMON);
                  ^

[With annotations:](https://github.com/lurtz/sleep-proxy/actions/runs/4213473474/jobs/7313201206)

    INFO:CPP Linter:Running "clang-tidy-14 --export-fixes=clang_tidy_output.yml -p /home/runner/work/sleep-proxy/sleep-proxy/build --extra-arg= src/emulateHost.cpp"
    DEBUG:CPP Linter:Output from clang-tidy:
    src/emulateHost.cpp:28:4: warning: unused variable 'x' [clang-diagnostic-unused-variable]
              x
              ^
    /home/runner/work/sleep-proxy/sleep-proxy/src/emulateHost.cpp:32:7: warning: redundant boolean literal in if statement condition [readability-simplify-boolean-expr]
      if (false) {
      ~~~~^~~~~~~~
    /home/runner/work/sleep-proxy/sleep-proxy/src/emulateHost.cpp:33:18: warning: variable 'sptr' is not initialized [cppcoreguidelines-init-variables]
              std::string * sptr;
                            ^
    note: this fix will not be applied because it overlaps with another fix
    /home/runner/work/sleep-proxy/sleep-proxy/src/emulateHost.cpp:40:15: warning: do not use pointer arithmetic [cppcoreguidelines-pro-bounds-pointer-arithmetic]
        setup_log(argv[0], 0, LOG_DAEMON);
                  ^

You can see in both cases absolute paths are used, but in the non working case a detour via "build/.." is taken.

--- a/.github/workflows/build-docs.yml
+++ b/.github/workflows/build-docs.yml
@@ -6,9 +6,9 @@ jobs:
   build:
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v3
+      - uses: actions/checkout@v4
 
-      - uses: actions/setup-python@v4
+      - uses: actions/setup-python@v5
         with:
           python-version: 3.x
 
@@ -19,7 +19,7 @@ jobs:
         run: sphinx-build docs docs/_build/html
 
       - name: upload docs build as artifact
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
           name: "cpp-linter_docs"
           path: ${{ github.workspace }}/docs/_build/html

--- a/.github/workflows/pre-commit-hooks.yml
+++ b/.github/workflows/pre-commit-hooks.yml
@@ -9,8 +9,8 @@ jobs:
   check-source-files:
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v3
-      - uses: actions/setup-python@v4
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
         with:
           python-version: '3.x'
       - run: python3 -m pip install pre-commit

--- a/.github/workflows/publish-pypi.yml
+++ b/.github/workflows/publish-pypi.yml
@@ -23,12 +23,12 @@ jobs:
     runs-on: ubuntu-latest
 
     steps:
-    - uses: actions/checkout@v3
+    - uses: actions/checkout@v4
       # use fetch --all for setuptools_scm to work
       with:
         fetch-depth: 0
     - name: Set up Python
-      uses: actions/setup-python@v4
+      uses: actions/setup-python@v5
       with:
         python-version: '3.x'
     - name: Install dependencies

--- a/.github/workflows/run-dev-tests.yml
+++ b/.github/workflows/run-dev-tests.yml
@@ -23,14 +23,14 @@ jobs:
   build:
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v3
-      - uses: actions/setup-python@v4
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
         with:
           python-version: '3.x'
       - name: Build wheel
         run: python3 -m pip wheel --no-deps -w dist .
       - name: Upload wheel as artifact
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
           name: cpp-linter_wheel
           path: ${{ github.workspace }}/dist/*.whl
@@ -46,14 +46,14 @@ jobs:
 
     runs-on: ${{ matrix.os }}
     steps:
-      - uses: actions/checkout@v3
+      - uses: actions/checkout@v4
 
-      - uses: actions/setup-python@v4
+      - uses: actions/setup-python@v5
         with:
           python-version: ${{ matrix.py }}
 
       - name: download wheel artifact
-        uses: actions/download-artifact@v3
+        uses: actions/download-artifact@v4
         with:
           name: cpp-linter_wheel
           path: dist
@@ -83,7 +83,7 @@ jobs:
           python -m pip install clang-tools
           clang-tools --install ${{ matrix.version }}
 
-      - name: Collect Coverage (native clang install)
+      - name: Collect Coverage
         env:
           CLANG_VERSION: ${{ matrix.version }}
         run: coverage run -m pytest

--- a/.github/workflows/run-dev-tests.yml
+++ b/.github/workflows/run-dev-tests.yml
@@ -96,10 +96,45 @@ jobs:
           CLANG_VERSION: ${{ matrix.version }}
         run: coverage run -m pytest
 
+      - name: Upload coverage data
+        uses: actions/upload-artifact@v4
+        with:
+          name: coverage-data-${{ runner.os }}-py${{ matrix.py }}-${{ matrix.version }}
+          path: .coverage*
+
+  coverage-report:
+    needs: [test]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: Download all artifacts
+        uses: actions/download-artifact@v4
+        with:
+          path: ci-artifacts
+
+      - run: mv ci-artifacts/**/.coverage* ./
+
+      - name: Setup python
+        uses: actions/setup-python@v5
+        with:
+          python-version: '3.x'
+
+      - name: Create coverage report
+        run: |
+          pip3 install coverage[toml]
+          coverage combine
+          coverage html
+
+      - name: Upload comprehensive coverage HTML report
+        uses: actions/upload-artifact@v4
+        with:
+          name: coverage-report
+          path: htmlcov/
+
       - run: coverage report && coverage xml
 
       - uses: codecov/codecov-action@v3
-        if: runner.os == 'Linux' && matrix.version == '15' && matrix.py == '3.10'
         with:
           token: ${{secrets.CODECOV_TOKEN}}
           files: ./coverage.xml

--- a/.github/workflows/run-dev-tests.yml
+++ b/.github/workflows/run-dev-tests.yml
@@ -63,6 +63,14 @@ jobs:
         shell: bash
         run: python3 -m pip install pytest coverage[toml] meson dist/*.whl
 
+      # https://github.com/ninja-build/ninja/wiki/Pre-built-Ninja-packages
+      - name: Install ninja (Linux)
+        if: runner.os == 'Linux'
+        run: sudo apt-get install ninja-build
+      - name: Install ninja (Windows)
+        if: runner.os == 'Windows'
+        run: choco install ninja
+
       - name: Install Linux clang dependencies
         if: runner.os == 'Linux'
         shell: bash

--- a/.github/workflows/run-dev-tests.yml
+++ b/.github/workflows/run-dev-tests.yml
@@ -40,23 +40,9 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        py: ['3.7', '3.8', '3.9', '3.10']
+        py: ['3.8', '3.9', '3.10', '3.11']
         os: ['windows-latest', ubuntu-22.04]
-        version: ['15', '14', '13', '12', '11', '10', '9', '8', '7']
-        include:
-          # mark all matrix jobs as "need to install clang-tools"
-          - tools_dir: 'temp'
-
-          # mark certain matrix jobs as "use natively installed clang-tools"
-          - os: 'windows-latest'
-            version: '14'
-            tools_dir: 'N/A'
-          - os: 'ubuntu-22.04'
-            version: '14'
-            tools_dir: 'N/A'
-          - os: 'ubuntu-22.04'
-            version: '13'
-            tools_dir: 'N/A'
+        version: ['17', '16', '15', '14', '13', '12', '11', '10', '9', '8', '7']
 
     runs-on: ${{ matrix.os }}
     steps:
@@ -77,28 +63,35 @@ jobs:
         shell: bash
         run: python3 -m pip install pytest coverage[toml] dist/*.whl
 
+      - name: Install Linux clang dependencies
+        if: runner.os == 'Linux'
+        shell: bash
+        run: |
+          sudo apt-get update
+          # First try installing from default Ubuntu repositories before trying LLVM script
+          if ! sudo apt-get install -y clang-format-${{ matrix.version }} clang-tidy-${{ matrix.version }}; then
+            # This LLVM script will add the relevant LLVM PPA: https://apt.llvm.org/
+            wget https://apt.llvm.org/llvm.sh -O ${{ runner.temp }}/llvm_install.sh
+            chmod +x ${{ runner.temp }}/llvm_install.sh
+            if sudo ${{ runner.temp }}/llvm_install.sh ${{ matrix.version }}; then
+              sudo apt-get install -y clang-format-${{ matrix.version }} clang-tidy-${{ matrix.version }}
+            fi
+          fi
+
       - name: Install clang-tools
-        if: matrix.tools_dir == 'temp'
         run: |
           python -m pip install clang-tools
-          clang-tools --install ${{ matrix.version }}  --directory ${{ runner.temp }}/clang-tools
+          clang-tools --install ${{ matrix.version }}
 
       - name: Collect Coverage (native clang install)
-        if: matrix.tools_dir == 'N/A'
         env:
           CLANG_VERSION: ${{ matrix.version }}
-        run: coverage run -m pytest
-
-      - name: Collect Coverage (non-native clang install)
-        if: matrix.tools_dir == 'temp'
-        env:
-          CLANG_VERSION: ${{ runner.temp }}/clang-tools
         run: coverage run -m pytest
 
       - run: coverage report && coverage xml
 
       - uses: codecov/codecov-action@v3
-        if: matrix.os == 'ubuntu-22.04' && matrix.version == '12' && matrix.py == '3.10'
+        if: runner.os == 'Linux' && matrix.version == '15' && matrix.py == '3.10'
         with:
           token: ${{secrets.CODECOV_TOKEN}}
           files: ./coverage.xml

--- a/.github/workflows/run-dev-tests.yml
+++ b/.github/workflows/run-dev-tests.yml
@@ -61,7 +61,7 @@ jobs:
       - name: Install workflow deps
         # using a wildcard as filename on Windows requires a bash shell
         shell: bash
-        run: python3 -m pip install pytest coverage[toml] dist/*.whl
+        run: python3 -m pip install pytest coverage[toml] meson dist/*.whl
 
       - name: Install Linux clang dependencies
         if: runner.os == 'Linux'

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -35,4 +35,4 @@ repos:
         types: [python]
         entry: mypy
         exclude: "^(docs/|setup.py$)"
-        additional_dependencies: [mypy, types-pyyaml, types-requests, rich, requests, pytest, pyyaml, '.']
+        additional_dependencies: [mypy, types-pyyaml, types-requests, rich, requests, pytest, pyyaml, meson, '.']

--- a/cpp_linter/clang_tidy.py
+++ b/cpp_linter/clang_tidy.py
@@ -2,7 +2,7 @@
 from pathlib import Path, PurePath
 from textwrap import indent
 import re
-from typing import Tuple, Union, List, cast
+from typing import Tuple, Union, List, cast, Optional, Dict
 from . import GlobalParser, CLANG_TIDY_STDOUT
 
 NOTE_HEADER = re.compile(r"^(.+):(\d+):(\d+):\s(\w+):(.*)\[([a-zA-Z\d\-\.]+)\]$")
@@ -20,6 +20,7 @@ class TidyNotification:
     def __init__(
         self,
         notification_line: Tuple[str, Union[int, str], Union[int, str], str, str, str],
+        database: Optional[List[Dict[str, str]]] = None,
     ):
         # logger.debug("Creating tidy note from line %s", notification_line)
         (
@@ -40,13 +41,32 @@ class TidyNotification:
         #: The line number of the source file.
         self.line = int(self.line)
         self.cols = int(self.cols)
-        #: The source filename concerning the notification.
-        self.filename = (
+        rel_path = (
             Path(self.filename)
             .resolve()
             .as_posix()
             .replace(Path.cwd().as_posix() + "/", "")
         )
+        if not PurePath(self.filename).is_absolute():
+            if database is not None:
+                # get absolute path from compilation database:
+                # This is need for meson builds as they use paths relative to
+                # the build env (or wherever the database is usually located).
+                for unit in database:
+                    if (
+                        "file" in unit
+                        and "directory" in unit
+                        and unit["file"] == self.filename
+                    ):
+                        rel_path = (
+                            Path(unit["directory"], unit["file"])
+                            .resolve()
+                            .as_posix()
+                            .replace(Path.cwd().as_posix() + "/", "")
+                        )
+                        break
+        #: The source filename concerning the notification.
+        self.filename = rel_path
         #: A `list` of lines for the code-block in the notification.
         self.fixit_lines: List[str] = []
 
@@ -93,7 +113,7 @@ class TidyNotification:
         )
 
 
-def parse_tidy_output() -> None:
+def parse_tidy_output(database: Optional[List[Dict[str, str]]]) -> None:
     """Parse clang-tidy output in a file created from stdout. The results are
     saved to :attr:`~cpp_linter.GlobalParser.tidy_notes`."""
     notification = None
@@ -105,7 +125,8 @@ def parse_tidy_output() -> None:
                 cast(
                     Tuple[str, Union[int, str], Union[int, str], str, str, str],
                     match.groups(),
-                )
+                ),
+                database,
             )
             GlobalParser.tidy_notes.append(notification)
         elif notification is not None:

--- a/cpp_linter/clang_tidy.py
+++ b/cpp_linter/clang_tidy.py
@@ -47,8 +47,7 @@ class TidyNotification:
             .as_posix()
             .replace(Path.cwd().as_posix() + "/", "")
         )
-        if not PurePath(self.filename).is_absolute():
-            if database is not None:
+        if not PurePath(self.filename).is_absolute() and database is not None:
                 # get absolute path from compilation database:
                 # This is need for meson builds as they use paths relative to
                 # the build env (or wherever the database is usually located).

--- a/cpp_linter/run.py
+++ b/cpp_linter/run.py
@@ -275,7 +275,6 @@ def run_clang_tidy(
     checks: str,
     lines_changed_only: int,
     database: str,
-    repo_root: str,
     extra_args: List[str],
 ) -> None:
     """Run clang-tidy on a certain file.
@@ -471,7 +470,6 @@ def capture_clang_tools_output(
             checks,
             lines_changed_only,
             database,
-            repo_root,
             extra_args,
         )
         run_clang_format(file, version, style, lines_changed_only)

--- a/cpp_linter/run.py
+++ b/cpp_linter/run.py
@@ -16,7 +16,7 @@ import configparser
 import json
 import urllib.parse
 import logging
-from typing import cast, List, Tuple, Optional
+from typing import cast, List, Tuple, Optional, Dict
 import requests
 from . import (
     Globals,
@@ -202,7 +202,7 @@ def filter_out_non_source_files(
         # dump altered json of changed files
         CHANGED_FILES_JSON.write_text(
             json.dumps(
-                [ f.serialize() for f in Globals.FILES ],
+                [f.serialize() for f in Globals.FILES],
                 indent=2,
             ),
             encoding="utf-8",
@@ -319,8 +319,6 @@ def run_clang_tidy(
         cmds.append(f"-checks={checks}")
     if database:
         cmds.append("-p")
-        if not PurePath(database).is_absolute():
-            database = str(Path(RUNNER_WORKSPACE, repo_root, database).resolve())
         cmds.append(database)
     line_ranges = {
         "name": filename,
@@ -391,6 +389,7 @@ def create_comment_body(
     file_obj: FileObj,
     lines_changed_only: int,
     tidy_notes: List[TidyNotification],
+    database: str,
 ):
     """Create the content for a thread comment about a certain file.
     This is a helper function to `capture_clang_tools_output()`.
@@ -401,9 +400,14 @@ def create_comment_body(
         avoid duplicated content in comment, and it is later used again by
         `make_annotations()` after `capture_clang_tools_output()` is finished.
     """
+    db_json: Optional[List[Dict[str, str]]] = None
+    if database:
+        db_json = json.loads(
+            Path(database, "compile_commands.json").read_text(encoding="utf-8")
+        )
     ranges = file_obj.range_of_changed_lines(lines_changed_only)
     if CLANG_TIDY_STDOUT.exists() and CLANG_TIDY_STDOUT.stat().st_size:
-        parse_tidy_output()  # get clang-tidy fixes from stdout
+        parse_tidy_output(db_json)  # get clang-tidy fixes from stdout
         comment_output = ""
         for fix in GlobalParser.tidy_notes:
             if lines_changed_only and fix.line not in ranges:
@@ -469,7 +473,7 @@ def capture_clang_tools_output(
         run_clang_format(file, version, style, lines_changed_only)
         end_log_group()
 
-        create_comment_body(file, lines_changed_only, tidy_notes)
+        create_comment_body(file, lines_changed_only, tidy_notes, database)
 
     if Globals.FORMAT_COMMENT or Globals.TIDY_COMMENT:
         Globals.OUTPUT += ":warning:\nSome files did not pass the configured checks!\n"
@@ -714,6 +718,10 @@ def main():
     else:
         list_source_files(args.extensions, ignored, not_ignored)
     end_log_group()
+
+    database: str = args.database
+    if not PurePath(database).is_absolute():
+        database = str(Path(RUNNER_WORKSPACE, args.repo_root, database).resolve())
 
     capture_clang_tools_output(
         args.version,

--- a/cpp_linter/run.py
+++ b/cpp_linter/run.py
@@ -310,10 +310,11 @@ def run_clang_tidy(
         CLANG_TIDY_STDOUT.write_bytes(b"")
         return
     filename = file_obj.name.replace("/", os.sep)
-    cmds = [
-        assemble_version_exec("clang-tidy", version),
-        f"--export-fixes={str(CLANG_TIDY_YML)}",
-    ]
+    cmds = [assemble_version_exec("clang-tidy", version)]
+    if "CPP_LINTER_TEST_ALPHA_CODE" in os.environ:
+        cmds.append(f"--export-fixes={str(CLANG_TIDY_YML)}")
+        # clear yml file's content before running clang-tidy
+        CLANG_TIDY_YML.write_bytes(b"")
     if checks:
         cmds.append(f"-checks={checks}")
     if database:
@@ -331,13 +332,11 @@ def run_clang_tidy(
     for extra_arg in extra_args:
         cmds.append(f"--extra-arg={extra_arg}")
     cmds.append(filename)
-    # clear yml file's content before running clang-tidy
-    CLANG_TIDY_YML.write_bytes(b"")
     logger.info('Running "%s"', " ".join(cmds))
     results = subprocess.run(cmds, capture_output=True)
     CLANG_TIDY_STDOUT.write_bytes(results.stdout)
     logger.debug("Output from clang-tidy:\n%s", results.stdout.decode())
-    if CLANG_TIDY_YML.stat().st_size:
+    if "CPP_LINTER_TEST_ALPHA_CODE" in os.environ and CLANG_TIDY_YML.stat().st_size:
         parse_tidy_suggestions_yml()  # get clang-tidy fixes from yml
     if results.stderr:
         logger.debug(

--- a/cpp_linter/run.py
+++ b/cpp_linter/run.py
@@ -389,7 +389,7 @@ def create_comment_body(
     file_obj: FileObj,
     lines_changed_only: int,
     tidy_notes: List[TidyNotification],
-    database: str,
+    database: Optional[List[Dict[str, str]]],
 ):
     """Create the content for a thread comment about a certain file.
     This is a helper function to `capture_clang_tools_output()`.
@@ -400,14 +400,9 @@ def create_comment_body(
         avoid duplicated content in comment, and it is later used again by
         `make_annotations()` after `capture_clang_tools_output()` is finished.
     """
-    db_json: Optional[List[Dict[str, str]]] = None
-    if database:
-        db_json = json.loads(
-            Path(database, "compile_commands.json").read_text(encoding="utf-8")
-        )
     ranges = file_obj.range_of_changed_lines(lines_changed_only)
     if CLANG_TIDY_STDOUT.exists() and CLANG_TIDY_STDOUT.stat().st_size:
-        parse_tidy_output(db_json)  # get clang-tidy fixes from stdout
+        parse_tidy_output(database)  # get clang-tidy fixes from stdout
         comment_output = ""
         for fix in GlobalParser.tidy_notes:
             if lines_changed_only and fix.line not in ranges:
@@ -457,6 +452,15 @@ def capture_clang_tools_output(
     :param extra_args: A list of extra arguments used by clang-tidy as compiler
         arguments.
     """
+
+    if database and not PurePath(database).is_absolute():
+        database = str(Path(RUNNER_WORKSPACE, repo_root, database).resolve())
+    db_json: Optional[List[Dict[str, str]]] = None
+    if database:
+        db_path = Path(database, "compile_commands.json")
+        if db_path.exists():
+            db_json = json.loads(db_path.read_text(encoding="utf-8"))
+
     # temporary cache of parsed notifications for use in log commands
     tidy_notes: List[TidyNotification] = []
     for file in Globals.FILES:
@@ -473,7 +477,7 @@ def capture_clang_tools_output(
         run_clang_format(file, version, style, lines_changed_only)
         end_log_group()
 
-        create_comment_body(file, lines_changed_only, tidy_notes, database)
+        create_comment_body(file, lines_changed_only, tidy_notes, db_json)
 
     if Globals.FORMAT_COMMENT or Globals.TIDY_COMMENT:
         Globals.OUTPUT += ":warning:\nSome files did not pass the configured checks!\n"
@@ -718,10 +722,6 @@ def main():
     else:
         list_source_files(args.extensions, ignored, not_ignored)
     end_log_group()
-
-    database: str = args.database
-    if not PurePath(database).is_absolute():
-        database = str(Path(RUNNER_WORKSPACE, args.repo_root, database).resolve())
 
     capture_clang_tools_output(
         args.version,

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -67,6 +67,9 @@ testpaths = ["tests"]
 [tool.coverage]
 [tool.coverage.run]
 dynamic_context = "test_function"
+# These options are useful if combining coverage data from multiple tested envs
+parallel = true
+relative_files = true
 omit = [
     # don't include tests in coverage
     "tests/*",

--- a/requirements-dev.txt
+++ b/requirements-dev.txt
@@ -1,4 +1,5 @@
 coverage[toml]
+meson
 mypy
 pre-commit
 pytest

--- a/tests/capture_tools_output/test_database_path.py
+++ b/tests/capture_tools_output/test_database_path.py
@@ -117,9 +117,13 @@ def test_ninja_database(
         repo_root=".",  # already changed cwd
         extra_args=[],
     )
+    found_project_file = False
     for note in cpp_linter.GlobalParser.tidy_notes:
         if note.filename.endswith("demo.cpp") or note.filename.endswith("demo.hpp"):
             assert not Path(note.filename).is_absolute()
+            found_project_file = True
+    if not found_project_file:
+        raise RuntimeError("no project files raised concerns with clang-tidy")
     assert make_annotations("", True, 0)
     # write step-summary for manual verification
     Path(tmp_path, "job_summary.md").write_text(

--- a/tests/capture_tools_output/test_database_path.py
+++ b/tests/capture_tools_output/test_database_path.py
@@ -96,7 +96,7 @@ def test_ninja_database(
     monkeypatch.setattr(
         cpp_linter.Globals,
         "FILES",
-        [FileObj("demo.cpp", [], []), FileObj("demo.hpp", [], [])],
+        [FileObj("demo.cpp", [], [])],
     )
     for attr in ["OUTPUT", "FORMAT_COMMENT", "TIDY_COMMENT"]:
         monkeypatch.setattr(cpp_linter.Globals, attr, "")

--- a/tests/capture_tools_output/test_database_path.py
+++ b/tests/capture_tools_output/test_database_path.py
@@ -4,11 +4,14 @@ from pathlib import Path, PurePath
 import logging
 import os
 import re
+import sys
+import shutil
 import pytest
 from cpp_linter import logger, FileObj
 import cpp_linter.run
 import cpp_linter
-from cpp_linter.run import capture_clang_tools_output
+from cpp_linter.run import capture_clang_tools_output, make_annotations
+from mesonbuild.mesonmain import main as meson  # type: ignore[import-untyped]
 
 CLANG_TIDY_COMMAND = re.compile(r'clang-tidy[^\s]*\s(.*)"')
 
@@ -66,3 +69,59 @@ def test_db_detection(
         raise RuntimeError("failed to find args passed in clang-tidy in log records")
     expected_args.append(demo_src.replace("/", os.sep) + '"')
     assert expected_args == matched_args
+
+
+def test_ninja_database(
+    monkeypatch: pytest.MonkeyPatch, tmp_path: Path, caplog: pytest.LogCaptureFixture
+):
+    """verify that the relative paths used in a database generated (and thus clang-tidy
+    stdout) for the ninja build system are resolved accordingly."""
+    tmp_path_demo = tmp_path / "demo"
+    # generate the project
+    shutil.copytree(
+        str(Path(__file__).parent.parent / "demo"),
+        str(tmp_path_demo),
+        ignore=shutil.ignore_patterns("compile_flags.txt"),
+    )
+    (tmp_path_demo / "build").mkdir(parents=True)
+    monkeypatch.chdir(str(tmp_path_demo))
+    monkeypatch.setattr(sys, "argv", ["meson", "init"])
+    meson()
+    monkeypatch.setattr(
+        sys, "argv", ["meson", "setup", "--backend=ninja", "build", "."]
+    )
+    meson()
+
+    caplog.set_level(logging.DEBUG, logger=cpp_linter.logger.name)
+    monkeypatch.setattr(
+        cpp_linter.Globals,
+        "FILES",
+        [FileObj("demo.cpp", [], []), FileObj("demo.hpp", [], [])],
+    )
+    for attr in ["OUTPUT", "FORMAT_COMMENT", "TIDY_COMMENT"]:
+        monkeypatch.setattr(cpp_linter.Globals, attr, "")
+    for attr in ["tidy_notes", "format_advice"]:
+        monkeypatch.setattr(cpp_linter.GlobalParser, attr, [])
+    for attr in ["tidy_failed_count", "format_failed_count"]:
+        monkeypatch.setattr(cpp_linter.Globals, attr, 0)
+    monkeypatch.setattr(cpp_linter.run, "RUNNER_WORKSPACE", str(tmp_path_demo))
+
+    cpp_linter.CLANG_TIDY_STDOUT.parent.mkdir(parents=True)
+    # run clang-tidy and verify paths of project files were matched with database paths
+    capture_clang_tools_output(
+        version=os.getenv("CLANG_VERSION", "12"),
+        checks="",  # let clang-tidy use a .clang-tidy config file
+        style="",  # don't invoke clang-format
+        lines_changed_only=0,  # analyze complete file
+        database="build",  # point to generated compile_commands.txt
+        repo_root=".",  # already changed cwd
+        extra_args=[],
+    )
+    for note in cpp_linter.GlobalParser.tidy_notes:
+        if note.filename.endswith("demo.cpp") or note.filename.endswith("demo.hpp"):
+            assert not Path(note.filename).is_absolute()
+    assert make_annotations("", True, 0)
+    # write step-summary for manual verification
+    Path(tmp_path, "job_summary.md").write_text(
+        cpp_linter.Globals.OUTPUT, encoding="utf-8"
+    )

--- a/tests/capture_tools_output/test_database_path.py
+++ b/tests/capture_tools_output/test_database_path.py
@@ -7,9 +7,10 @@ import re
 import pytest
 from cpp_linter import logger, FileObj
 import cpp_linter.run
-from cpp_linter.run import run_clang_tidy
+import cpp_linter
+from cpp_linter.run import capture_clang_tools_output
 
-CLANG_TIDY_COMMAND = re.compile(r'"clang-tidy(.*)"')
+CLANG_TIDY_COMMAND = re.compile(r'clang-tidy[^\s]*\s(.*)"')
 
 ABS_DB_PATH = str(Path("tests/demo").resolve())
 
@@ -20,7 +21,7 @@ ABS_DB_PATH = str(Path("tests/demo").resolve())
         # implicit path to the compilation database
         ("", []),
         # explicit relative path to the compilation database
-        ("../../demo", ["-p", ABS_DB_PATH]),
+        ("demo", ["-p", ABS_DB_PATH]),
         # explicit absolute path to the compilation database
         (ABS_DB_PATH, ["-p", ABS_DB_PATH]),
     ],
@@ -37,25 +38,31 @@ def test_db_detection(
     monkeypatch.chdir(PurePath(__file__).parent.as_posix())
     cpp_linter.CACHE_PATH.mkdir(exist_ok=True)
     demo_src = "../demo/demo.cpp"
-    rel_root = str(Path(*Path(__file__).parts[-2:]))
-    cpp_linter.run.RUNNER_WORKSPACE = (
-        Path(pytestconfig.rootpath, "tests").resolve().as_posix()
+    monkeypatch.setattr(
+        cpp_linter.run,
+        "RUNNER_WORKSPACE",
+        Path(pytestconfig.rootpath, "tests").resolve().as_posix(),
     )
+    monkeypatch.setattr(cpp_linter.Globals, "FILES", [FileObj(demo_src, [], [])])
+
     caplog.set_level(logging.DEBUG, logger=logger.name)
-    run_clang_tidy(
-        file_obj=FileObj(demo_src, [], []),  # not filtering lines
-        version="",
+    capture_clang_tools_output(
+        version=os.getenv("CLANG_VERSION", "12"),
         checks="",  # let clang-tidy use a .clang-tidy config file
+        style="",  # don't invoke clang-format
         lines_changed_only=0,  # analyze complete file
         database=database,
-        repo_root=rel_root,
+        repo_root=".",
         extra_args=[],
     )
     matched_args = []
     for record in caplog.records:
+        assert "Error while trying to load a compilation database" not in record.message
         msg_match = CLANG_TIDY_COMMAND.search(record.message)
         if msg_match is not None:
-            matched_args = msg_match.group(0)[:-1].split()[2:]
-        assert "Error while trying to load a compilation database" not in record.message
-    expected_args.append(demo_src.replace("/", os.sep))
-    assert matched_args == expected_args
+            matched_args = msg_match.group(0).split()[2:]
+            break
+    else:
+        raise RuntimeError("failed to find args passed in clang-tidy in log records")
+    expected_args.append(demo_src.replace("/", os.sep) + '"')
+    assert expected_args == matched_args

--- a/tests/capture_tools_output/test_database_path.py
+++ b/tests/capture_tools_output/test_database_path.py
@@ -60,7 +60,7 @@ def test_db_detection(
         assert "Error while trying to load a compilation database" not in record.message
         msg_match = CLANG_TIDY_COMMAND.search(record.message)
         if msg_match is not None:
-            matched_args = msg_match.group(0).split()[2:]
+            matched_args = msg_match.group(0).split()[1:]
             break
     else:
         raise RuntimeError("failed to find args passed in clang-tidy in log records")

--- a/tests/capture_tools_output/test_tools_output.py
+++ b/tests/capture_tools_output/test_tools_output.py
@@ -352,6 +352,7 @@ def test_diff_comment(
     comments (not annotations) in the event's diff.
 
     Remember, diff comments should only focus on lines in the diff."""
+    monkeypatch.setenv("CPP_LINTER_TEST_ALPHA_CODE", "true")
     prep_tmp_dir(
         tmp_path,
         monkeypatch,

